### DISCUSSION TITLE
Extension to open a browser tab inside code-server

### DISCUSSION THREAD

**Discussion by metalshanked** [coder/code-server#3318] [General]
Hi, I have code-server running on a docker container and it can access specific sites in its VLAN.
So was wondering if there is an extension anyone knows about which can open a browser tab for accessing webpages directly from inside code-server?

I installed chrome inside the code-server container and i have tried the only extension i could find --> https://marketplace.visualstudio.com/items?itemName=auchenberg.vscode-browser-preview but it is a bit clunky and does not open pages well.

Here is what it should look like (based on above extension i tried)
![image](https://user-images.githubusercontent.com/57646596/117505527-0366d680-af39-11eb-8c72-4a2392cba340.png).

**Comment by jsjoeio** [comment]
I don't know of any off the top of my head 🤔 There were a couple similar discussions here:
- https://github.com/cdr/code-server/discussions/3128
- https://github.com/cdr/code-server/discussions/3088

**Reply by metalshanked** [reply]
Thanks @jsjoeio .  My use case is slightly different as in I was wondering if there is a "general purpose" browser available within code-server to browser websites (Eg:- external URLs). I have google-chrome installed in the code-server instance (Docker)
The only extension I could find was [this one ](https://marketplace.visualstudio.com/items?itemName=auchenberg.vscode-browser-preview) but it does not work well.

So basically i am looking for a browser UI extension or feature which can interact with the installed  chrome binary "headlessly" and display the webpages entered in the URL bar?

**Reply by jsjoeio** [reply]
Ah, I see. I've looked for something like this in the past but haven't had much luck either 😢 Maybe @bpmct knows of something? He seems to be a wizard when it comes to extensions.

**Reply by bpmct** [reply]
> The only extension I could find was this one but it does not work well.

What problems to you run into with the Browser Preview extension? It is possible that it has been fixed in the latest version, which is not in the code-server marketplace.

<img width="1569" alt="Screen Shot 2021-05-13 at 4 41 45 PM" src="https://user-images.githubusercontent.com/22407953/118185647-c2743380-b40a-11eb-8992-19fa38327409.png">

I built the latest version of the browser preview extension from the GitHub and installed from VSIX and it seemed to work great!
```sh
git clone https://github.com/auchenberg/vscode-browser-preview.git
cd vscode-browser-preview
npm install

# build the VSIX
npm install -g vsce
vsce package

# in the extensions panel, press the ... and click "Install from VSIX." Select the file
```

**Reply by bpmct** [reply]
You may need Google chrome / chromium installed on your host machine to get it to work well, it maybe using a bad browser if it doesn't have chromium.

**Reply by metalshanked** [reply]
Thanks @bpmct  - I believe i used the latest version from the marketplace. In general, kudos to the dev for putting this unique extension which no other seem to fulfil.  
The issues i face is that it is generally clunky and blurry and sometime i have to click the icon twice or thrice to have the page open.
I tried both 1) chromium via pupeteer via npm and 2) the official google-chrome-stable.deb as the browser engine with same results.

Perhaps it may be my config but my experience was in line with the majority of comments/reviews for the extension.

So was wondering if there was any option# 2 out there in the wild...

**Reply by bpmct** [reply]
> I believe i used the latest version from the marketplace

From the code-server marketplace or the VS Code one? The one in code-server is out of date

hmm.. nothing comes to mind besides Browser Preview. Is there a site in particular browser preview struggles with? Would like to test on my end.

**Reply by metalshanked** [reply]
from the vscode one --> https://marketplace.visualstudio.com/items?itemName=auchenberg.vscode-browser-preview

Thanks, my code-server is running as container and the use case was to use the plugin to access internal websites only available on the container network space.

**Reply by bpmct** [reply]
Ah I meant the URLs specifically, but given they are internal I see how it can't be shared 😅

the extension has worked without bugs for me. have you tried running code-server locally + Broswer preview instead of running as a container?

**Reply by metalshanked** [reply]
Oh ok, never tried that. It was more of a nice to have and technically works so did not try too hard for alternative installs.
Dockerized code-server works great
